File: wiseSaying/src/main/java/wiseSayingRepository/WiseSayingRepository.java
package wiseSayingRepository;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSyntaxException;
import data.Proverb;

import java.io.*;
import java.nio.Buffer;
import java.util.ArrayList;
import java.util.List;

public class WiseSayingRepository {
    private static final String DB = "src/main/java/db/";
    private static final String LAST_ID_FILE = DB + "lastId.txt";
    private static final String DATA_PROVERB_FILE = DB + "data.json";

    public int readLastId() {
        File file = new File(LAST_ID_FILE);
        if (!file.exists()) {
            try {
                if (file.createNewFile()) {
                    System.out.println("파일이 생성되었습니다: " + file.getAbsolutePath());
                } else {
                    System.out.println("파일 생성에 실패했습니다.");
                }
                return 1;

            } catch (IOException e) {
                System.out.println("파일 생성 중 오류 발생: " + e.getMessage());
            }
        } else {
            try (BufferedReader br = new BufferedReader(new FileReader(file))) {
                return Integer.parseInt(br.readLine());
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return 1;
    }

    public void saveLastId(int id) {
        try {
            BufferedWriter bw = new BufferedWriter(new FileWriter(LAST_ID_FILE));
            bw.write(String.valueOf(id));
            bw.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void saveProverb(Proverb proverbObject) {
        Gson gson = new Gson();
        String json = gson.toJson(proverbObject);

        try {
            FileWriter writer = new FileWriter(DB + proverbObject.getId() + ".json");
            writer.write(json);
            writer.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }


    }

    public List<Proverb> loadProverbs() throws FileNotFoundException {
        File folder = new File(DB);
        List<Proverb> proverbs = new ArrayList<>();

        File[] files = folder.listFiles(((dir, name) -> name.endsWith(".json")));
        if (files != null) {
            Gson gson = new Gson();

            for (File file : files) {
                try (BufferedReader br = new BufferedReader(new FileReader(file))) {
                    Proverb proverb = gson.fromJson(br, Proverb.class);
                    proverbs.add(proverb);
                } catch (JsonSyntaxException e) {
                    System.out.println("JSON 파싱 오류가 발생했습니다: " + file.getName());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return proverbs;
    }

    public void deleteProverbFile(int id) {
        File file = new File(DB + id + ".json");
        if (file.exists()) {
            if (file.delete()) {
                System.out.println(id + ".json 파일이 성공적으로 삭제되었습니다.");
            } else {
                System.out.println(id + ".json 파일 삭제에 실패하였습니다.");
            }
        } else {
            System.out.println(id + ".json 파일이 존재하지 않습니다.");
        }
    }

    public void saveDataToFile(List<Proverb> proverbList) {
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        // String json = gson.toJson(proverbList);
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(DATA_PROVERB_FILE))) {
            gson.toJson(proverbList, bw);
            System.out.println("빌드 성공");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: migrationWiseSaying/src/main/kotlin/repository/WiseSayingRepository.kt
package repository

import com.google.gson.Gson
import proverb.Proverb
import util.ExceptionHandler
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException

class WiseSayingRepository {

    // static
    companion object {
        private val DB = "src/main/kotlin/db"
        private val LAST_ID_FILE = DB + "lastId.txt"
        private val DATA_PROVERB_FILE = DB + "data.json"
    }

    fun readLastId(): Int {
        val file = File(LAST_ID_FILE)
        if (!file.exists()) {
            try {
                when {
                    file.createNewFile() -> println("파일이 생성되었습니다: ${file.absolutePath}")
                    else -> println("파일 생성에 실패했습니다.")
                }
                return 1
            } catch (e: IOException) {
                ExceptionHandler.handleIOException(e, "lastId.txt 파일 생성 중 오류 발생")
            }
        } else {
            try {
                BufferedReader(FileReader(file)).use { br -> // use 는 Closeable 리소스를 자동으로 닫아준다.,  Closeable 리소스란 입출력 스트림이나 소켓 등을 말한다.
                    return br.readLine().toInt() // .toInt() 는 문자열을 정수로 변환한다. = Integer.parseInt (in Java)
                }
            } catch (e: IOException) {
                ExceptionHandler.handleIOException(e, "lastId.txt 파일 읽기 중 오류 발생")
            }
        }
        return 1
    }

    fun saveProverb(proverbObject: Proverb) {
        val gson: Gson = Gson()
        var json: String = gson.toJson(proverbObject)

        try {
            /*val writer: FileWriter = FileWriter(DB + proverbObject.id +  ".json")
            writer.write(json)*/
            FileWriter(DB + "${proverbObject.id}.json").use { writer ->
                writer.write(json)
            }
        } catch (e: IOException) {
            ExceptionHandler.handleIOException(e, ".json 파일 저장 중 오류 발생")
        }
    }

    fun saveLastId(id: Int) {
        try {
            BufferedWriter(FileWriter(LAST_ID_FILE)).use {
                it.write(id.toString())
            }
        } catch (e: IOException) {
            ExceptionHandler.handleIOException(e, "lastId.txt  파일 쓰기 중 오류 발생")
        }
    }

    fun saveDataToFile(proverbList: MutableList<Proverb>) {
        val gson: Gson = Gson()
        try {
            BufferedWriter(FileWriter(DATA_PROVERB_FILE)).use { bw ->
                gson.toJson(proverbList, bw)
                println("빌드 성공")
            }
        } catch (e: IOException) {
            ExceptionHandler.handleIOException(e, "data.json 파일 쓰기 중 오류 발생")
        }
    }

    fun deleteProverbFile(id: Int) {
        val file = File("$DB$id.json")
        when {
            !file.exists() -> println("$id.json 파일이 존재하지 않습니다.")
            file.delete() -> println("$id.json 파일이 성공적으로 삭제되었습니다.")
            else -> println("$id.json 파일 삭제에 실패하였습니다.")
        }
    }
}